sled-agent/src/sim/storage.rs

```diff
@@ -1009,7 +1099,149 @@ impl Storage {
+                return Ok(children);
+            }
+        }
+    }
+
+    pub fn nested_dataset_ensure(
+        &mut self,
+        config: NestedDatasetConfig,
+    ) -> Result<(), HttpError> {
+        let name = &config.name;
+        let nested_path = name.path.to_string();
+        let zpool_root = self.root().to_path_buf();
+        let Some(mut nested_dataset) = self.nested_datasets.get_mut(&name.root)
+        else {
+            return Err(HttpError::for_not_found(
+                None,
+                "Dataset not found".to_string(),
+            ));
+        };
+
+        for path_component in nested_path.split('/') {
+            if path_component.is_empty() {
+                continue;
+            }
+
+            // Final component of path -- insert it here if it doesn't exist
+            // already.
+            if !path_component.contains('/') {
+                let entry =
+                    nested_dataset.children.entry(path_component.to_string());
+                entry
+                    .and_modify(|storage| {
+                        storage.config = config.clone();
+                    })
+                    .or_insert_with(|| {
+                        NestedDatasetStorage::new(
+                            &zpool_root,
+                            config.name.root,
+                            nested_path,
+                            config.inner,
+                        )
+                    });
+                return Ok(());
+            }
+
+            match nested_dataset.children.get_mut(path_component) {
+                Some(dataset) => nested_dataset = dataset,
+                None => {
+                    return Err(HttpError::for_not_found(
+                        None,
+                        "Dataset not found".to_string(),
+                    ))
+                }
+            };
+        }
+        return Err(HttpError::for_not_found(
+            None,
+            "Nested Dataset not found".to_string(),
+        ));
+    }
+
+    pub fn nested_dataset_destroy(
+        &mut self,
+        name: NestedDatasetLocation,
+    ) -> Result<(), HttpError> {
+        let Some(mut nested_dataset) = self.nested_datasets.get_mut(&name.root)
+        else {
+            return Err(HttpError::for_not_found(
+                None,
+                "Dataset not found".to_string(),
+            ));
+        };
+
+        for path_component in name.path.split('/') {
+            if path_component.is_empty() {
+                continue;
+            }
+
+            // Final component of path -- remove it if it exists.
+            if !path_component.contains('/') {
+                if nested_dataset.children.remove(path_component).is_none() {
+                    return Err(HttpError::for_not_found(
+                        None,
+                        "Nested Dataset not found".to_string(),
+                    ));
+                };
+                return Ok(());
+            }
```

Comment:
As above, no component will contain `/`, so we'll always exit early and fail to iterate past the first component.